Best positioning of lxqt-runner with multi-screen Wayland

tsujan opened the pull request:
There is a detailed comment in the code, but in short,

 * There is no Qt way for knowing the Wayland screen with the cursor. As a result, the size and vertical position of lxqt-runner (in case it's supposed to be centered vertically) are calculated and correct only relative to the smallest screen (which is usually the one with the largest scale factor). This means that the widget is wider than expected and above the center of another screen.
 * Also, https://github.com/lxqt/lxqt-runner/commit/b0cad7975f5806bcd238f2d4285b5998a9857fb7 is reversed because, at least on KWin, the screen of lxqt-runner doesn't change without a hidden parent — don't ask me why, but perhaps the X11 developers encountered a similar issue.

NOTE: So far, this compromise is the only way — until Qt provides better Wayland methods. Take it or leave it ;)

stefonarch reviewed (commented):
Here there is still the issue which can be seen when hitting the shortcut twice: the firs window is too large, the second one correct but too much on the left.

https://github.com/user-attachments/assets/6503f9f6-4937-4899-aa14-358d31fbc551

That is on the laptop monitor when the external is scaled up to 1.5.

stefonarch commented:
> > Here there is still part of the issue
> 
> I can't reproduce it here with labwc and kwin_wayland. You mean on Niri or in general?

I thought it was a Niri issue for some time, but I can reproduce the same on labwc:

1. use 1.1 scale for laptop, external on the right
2. open runner on the external and press twice
3. with external scale=1.5 issue is more evident on the laptop
4. Even with both scaled to identical values  a very small change can be seen.

tsujan commented:
> Maybe we should use px instead of percentage for the width?

Remember https://github.com/lxqt/lxqt-runner/commit/4022c9d97b3857150591d7089b9d708a5b750d31? Percentage was used because it prevented window resizing.

I'll try to reproduce what you described…

tsujan commented:
With the external monitor scaled to 1.5, I can reproduce it under KWin by pressing the shortcut twice. The reason that I hadn't seen it was that I didn't have a reason to press the shortcut twice ;) Will look into it.

tsujan commented:
I found the reason, added a workaround, and simplified some lines. I also added a short story as the code comment.

It passed my tests, but please be cruel when testing.

tsujan commented:
BTW, with the above changes, it seems easy to make the shortcut work on Wayland as it does on X11, i.e., to make the second shortcut press hide the window. But it'll be an experiment for another PR …

stefonarch reviewed (commented):
The issue when opening tiwce has gone, but the width is way off on the internal screen when the external is set to 1.5 scale. Also if the internal is set to scale 1.1 20% width results in 1/3 of screen on the external.

tsujan commented:
Please see if that's specific to Niri because I can't reproduce it here.

tsujan commented:
Oh, wait. you're talking about the width, not the horizontal position.

The width is correct. It's calculated based on the smallest screen.

stefonarch commented:
Width= 50%, HDMI-A-1 1.5 scale, Labwc:

![runner](https://github.com/user-attachments/assets/1cfd390c-25da-45a0-8c68-578435b6f3ca)

width=0%/disabled/400px minimum
![runner0](https://github.com/user-attachments/assets/2cd50976-3e96-4232-a2e7-1dfdc2d65ea0)

It works acceptable for values between 80 and 100%.

tsujan commented:
It's correct. I explained it in both PR comment and code comment. It's exactly like the case of Leave dialog, except it's about width.

stefonarch commented:
It may be correct from your point of view, but not for an user which sees it eventually superlarge. Maybe we should just remove width settings or propose 3 fixed values, 400, 500 600? 
And it should be tested with a real HDPi 4K Monitor. We could also merge that and wait for feedbacks/issues.

tsujan commented:
> It may be correct from your point of view

It's correct from the point of view of the smallest screen. As I said, here "small" doesn't mean the real size or even resolution. It means the size that's reported by the screen. The greater the scale factor is, the smaller the screen will be.

Nothing is left to chance.

> And it should be tested with a real HDPi 4K Monitor.

All tests can be done with ordinary monitors by using a great scale factor. But a scale factor > 2 is somehow imaginary ;)

tsujan commented:
> Maybe we should just remove width settings or propose 3 fixed values, 400, 500 600?

Why?! That would be a weird regression :o  There isn't the slightest reason for doing so — to say nothing about the case of a single monitor, or two monitors with equal sizes and without scale factors.

stefonarch reviewed (approved):
Let's see...

tsujan commented:
This is the same discussion we had before :) You have two monitors with the same resolution. In the real world, one applies a large scale factor to an external monitor when its resolution is much higher than the internal monitor, such that, in the end, the screen sizes become almost equal.

If I've learned something from wrestling with problems of multi-screen scaling on Wayland, it's that the window size shouldn't be touched (because the screen is usually wrong inside `showEvent`); instead, the layer margins should be set, by consulting the smallest screen. As a result, the window size on larger screens (whose corresponding monitors may be smaller) may not be what's expected. As I said in the PR comment, this is a compromise — the best one possible.
